plugins/pipelines/app/models/concerns/samson_pipelines/stage_concern.rb

```diff
@@ -10,6 +12,27 @@ def next_stages
 
   protected
 
+  # Needs to find all the possible stages in case this is a pipeline of pipelines as each subsequent stage
+  # could have valid next_stage_ids
+  def all_next_stages
+    existing_stages = next_stages.collect(&:id).to_set
+    return [] if existing_stages.empty?
+    loop do
+      last_count = existing_stages.count
+      existing_stages.each do |stage_id|
+        existing_stages += Stage.find(stage_id).next_stage_ids.map(&:to_i).to_set
+      end
+      break if last_count == existing_stages.count
+    end
+    # if we find ourselves with ourselves in the list, we have a loop.
+    # this shouldn't happen, but this will prevent it
+    if existing_stages.include?(id)
```

Comment:
Honestly I'd like to just to back to the recursive code that was there and rely on the model validation to prevent a loop (as it was before).

thoughts?